src/ZODB/mvccadapter.py

```diff
@@ -142,8 +131,15 @@ def sync(self, force=True):
             self._sync()
 
     def poll_invalidations(self):
+        # Storage implementations don't always call invalidate() when
+        # the last TID changes, e.g. after network reconnection,
+        # so we still have to poll.
+        ltid = self._storage.lastTransaction()
+        # But at this precise moment, a transaction may be committed and
+        # we have already received the new tid, along with invalidations.
```

Comment:
@jmuchemb, @jamadden isn't the above true only if storage, upon receiving invalidation messages from server, first notifies `DB.invalidate()` and only _after_ updates `zstor.last_tid`? For the reference this is indeed true for ZEO4:

https://github.com/zopefoundation/ZEO/blob/47d3fbe8cbf24cad91b183483df069ef20708874/src/ZEO/ClientStorage.py#L1473-L1482

but not true for NEO:

https://lab.nexedi.com/nexedi/neoppod/blob/fc58c089/neo/client/handlers/master.py#L109-124

If `zstor.last_tid` is updated first *before* `zstor` sends invalidations to DB (to MVCCAdapter here), what could happen is:

- poll_invalidations calls `zstor.lastTransaction` -> gets just committed  tid;
- poll invalidations proceeds and could retrieve `._invalidations` **before** `zstor` could actually update them with invalidation data corresponding to `tid`;
- tid, and incomplete set of objects to invalidate is returned.

I.e. it will be a data corruption, because objects might remain stale in live cache with similar data corruption scenario as demonstrated in https://github.com/zopefoundation/ZODB/issues/290.

P.S. I could not quickly see which way ZEO5 implements. Maybe it is something different, or actually related to https://github.com/zopefoundation/ZEO/issues/155.

P.P.S RelStorage is unaffected as it implements IMVCCStorage natively.

Reply:
There's indeed a problem between this MR & NEO :(

For ZEO4, the 2 operations are wrapped around the same lock so the order does not matter.

For ZEO5, it's first `DB.invalidate()` then lastTransaction() updated.

So I have to answer 2 questions:
1. Is ZEO5 behaviour enough ?
2. If it is, can I do the same in NEO ?

zopefoundation/ZEO#155 is unrelated.

Reply:
@jmuchemb, thanks for feedback.

Yes, I believe if we change `DB.invalidate()` call to always happen before `zstor.last_tid` update that would be enough. If we go this way this change also **_requires corresponding change in IStorage interface definiton_**. Something like below:

```diff
--- a/src/ZODB/interfaces.py
+++ b/src/ZODB/interfaces.py
@@ -685,6 +685,15 @@ def isReadOnly():
     def lastTransaction():
         """Return the id of the last committed transaction.
 
+        Returned tid is ID of last committed transaction as observed from
+        some time _before_ lastTransaction call was made. In particular for
+        client-sever case, lastTransaction can return cached view of storage
+        that was learned some time ago.
+
+        It is guaranteed that for all IStorageWrappers, that wrap the storage,
+        invalidation notifications have been completed for transactions
+        with ID ≤ returned tid.
+
         If no transactions have been committed, return a string of 8
         null (0) characters.
         """
```

Reply:
Indeed, I found no way to cause any bug if lastTransaction() is updated just after invalidations are notified. I feared issues because NEO also uses lastTransaction() internally: note however it must do it before releasing its cache lock, otherwise "load current" fails on ZODB<5.

Updating lastTransaction() at the right moment also matters for "internal" invalidations: I mean that it must be done just after calling the tpc_finish callback.

https://lab.nexedi.com/nexedi/neoppod/commit/5870c5d79f68dce00267f3ae9c4e12d56d19dede should make NEO work with this MR. It includes tests for the 4 cases.

Reply:
@jmuchemb, thanks for feedback and adapting things on NEO side. I see about the need to update `app.last_tid` while still under `app._cache_lock` - as `app._cache` and `app.last_tid` need to be kept coherent on NEO side internally.

> Updating lastTransaction() at the right moment also matters for "internal" invalidations: I mean that it must be done just after calling the tpc_finish callback.

Yes, thanks for pointing this out. There `app.last_tid` has to be also set while still under `app._cache_lock` and after call to `app._cache.invalidate()`. As you say this matters for "load current" on ZODB<5. I thus suggest the following amendment for ZODB4 version of proposed [IStorage.lastTransaction interface change](https://github.com/zopefoundation/ZODB/pull/307#discussion_r434391715) (ZODB4 specifics in bold):

> def lastTransaction():
> """Return the id of the last committed transaction.
> 
> *Returned tid is ID of last committed transaction as observed from*
> *some time _before_ lastTransaction call was made. In particular for*
> *client-sever case, lastTransaction can return cached view of storage*
> *that was learned some time ago.*
>
> *It is guaranteed that for all IStorageWrappers, that wrap the storage,*
> *invalidation notifications have been completed for transactions*
> *with ID ≤ returned tid.*
>
> **It is guaranteed that after lastTransaction returns, "current" view of**
> **the storage as observed by load() is ≥ returned tid.**
>
> If no transactions have been committed, return a string of 8
> null (0) characters.

I believe *it is very important we don't just fixup NEO and merge this patch without fixing IStorage interface*. There are application-level cases that rely on exactly the same guarantees that MVCCAdapter needs here and got wrong in the first version of this patch. Also it has to be clear for all other potential storages and their authors - just from reading the interfaces - which semantic must be provided.

I created the following pull-request to adjust the IStorage interface:

https://github.com/zopefoundation/ZODB/pull/313
https://github.com/zopefoundation/ZODB/pull/314

/cc @jamadden, @vpelletier, @arnaud-fontaine, @jwolf083, @klawlf82, @gitzit, @jimfulton

Reply:
> I believe _it is very important we don't just fixup NEO and merge this patch without fixing IStorage interface_.

Yes. First, I wanted to be sure this MR is correct.

> as observed from some time _before_ lastTransaction call was made.

Strictly speaking, it's wrong because the value that lastTransaction reads could be updated between the call to lastTransaction start and the instruction that actually reads the value in question.

I suggest the following change:
```diff
--- a/src/ZODB/interfaces.py
+++ b/src/ZODB/interfaces.py
@@ -685,6 +685,14 @@ def isReadOnly():
     def lastTransaction():
         """Return the id of the last committed transaction.
 
+        For proper MVCC operation, the return value is the id of the last
+        transaction for which invalidation notifications are completed.
+
+        In particular for client-server implementations, lastTransaction
+        should return a cached value (rather than querying the server).
+        A preliminary call to sync() can be done to get the actual last
+        TID at the wanted time.
+
         If no transactions have been committed, return a string of 8
         null (0) characters.
         """
```

Reply:
@jmuchemb, thanks for feedback.

> Strictly speaking, it's wrong ...

You are right, thanks for noticing it. It was a thinko on my side that I copied from [ZODB/go](https://godoc.org/lab.nexedi.com/kirr/neo/go/zodb#IStorage). The correction is to require _"as observed from some time *before* lastTransaction call ~was made~**returns**."_ I've [amended](https://lab.nexedi.com/kirr/neo/commit/949d42da) ZODB/go and my two pull requests accordingly.

> I suggest the following change:

I prefer my version, but am ok if `ZODB/interfaces.py` is updated with your text, or any other text that clearly documents required semantic.